plugins/analytics-module-ga4/src/util/DeferredCapture.ts

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import ReactGA from 'react-ga4';
+
+type Hit = {
+  timestamp: number;
+  data: {
+    hitType: 'pageview' | 'event';
+    [x: string]: any;
+  };
+};
+
+/**
+ * A wrapper around ReactGA that can optionally handle latent capture logic.
+ *
+ * - When defer is `false`, event data is sent directly to GA.
+ * - When defer is `true`, event data is queued (with a timestamp), so that it
+ *   can be sent to GA once externally indicated to be ready. This relies on
+ *   the `qt` or `queueTime` parameter of the Measurement Protocol.
+ *
+ * @see https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters#qt
+ */
+export class DeferredCapture {
+  /**
+   * Queue of deferred hits to be processed when ready. When undefined, hits
+   * can safely be sent without delay.
+   */
+  private queue: Hit[] | undefined;
+
+  constructor({ defer = false }: { defer: boolean }) {
+    this.queue = defer ? [] : undefined;
+  }
+
+  /**
+   * Indicates that deferred capture may now proceed.
+   */
+  setReady() {
+    if (this.queue) {
+      this.queue.forEach(this.sendDeferred);
+      this.queue = undefined;
+    }
+  }
+
+  /**
+   * Either forwards the pageview directly to GA, or (if configured) enqueues
+   * the pageview hit to be captured when ready.
+   */
+  pageview(path: string, metadata: any = {}) {
+    if (this.queue) {
+      this.queue.push({
+        timestamp: Date.now(),
+        data: {
+          hitType: 'pageview',
+          page: path,
+          ...metadata,
+        },
+      });
+      return;
+    }
+
+    ReactGA.send({
+      hitType: 'pageview',
+      page: path,
+      ...metadata,
+    });
+  }
+
+  /**
+   * Either forwards the event directly to GA, or (if configured) enqueues the
+   * event hit to be captured when ready.
+   */
+  event(eventDetails: any) {
+    if (this.queue) {
+      this.queue.push({
+        timestamp: Date.now(),
+        data: {
+          ...eventDetails,
+          hitType: 'event',
+        },
+      });
+      return;
+    }
+
+    ReactGA.event(eventDetails);
+  }
+
+  /**
+   * Sends a given hit to GA, decorated with the correct queue time.
+   */
+  private sendDeferred(hit: Hit) {
+    // Send the hit with the appropriate queue time (`qt`).
+    ReactGA.send({
+      ...hit.data,
+      queueTime: Date.now() - hit.timestamp,
+    });
+  }
```

Comment:
In conversations with folks [in the `#analytics` channel in discord](https://discord.com/channels/687207715902193673/1053066591891378206/1068204690145288292), I've heard that GA4 doesn't have an equivalent concept to queue time (`qt`).  The closest I could find is [`timestamp_micros` on the measurement protocol references](https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference?client_type=firebase#reserved_parameter_names) (which are not 100% the same / relevant for web-based tracking)

Are you sure this is working?

Reply:
GA folks in discord mentioned that queue time has been replaced and as you had mentioned timestamp_micros is the way to send events for the past.
From Google Analytics 
```
You can send the timestamp in GA4, so queue time shouldn't be needed: https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference?client_type=firebase#payload_post_body
```
That being said, I couldn't find a concrete way to test this. But, I could see that timestamp_micros was populated when I enabled debug in the GA4.

Reply:
Is removing deferredCapture an option?

Reply:
If we can maintain it using `timestamp_micros`, that would be ideal!

I think the best way to test is to...

1. configure the plugin to require an identity
2. Load backstage in the browser, fresh (signed out)
3. Wait 1 minute
4. Sign in and navigate to a different page

In the real-time view in GA, you should see the initial navigation separate by the amount of time you waited from the subsequent navigation event

Reply:
The other PR uses timestamp_micros, but yeah I will see if can test this proper. https://github.com/backstage/backstage/pull/16642/files#diff-d50aa3f56b50a43960d55c8b023e7870606424ce10b39dd522b499b1f933f89eR86